### DISCUSSION TITLE
Setting TenantId For New Users in Serene .net core

### DISCUSSION THREAD

**Discussion by imhemayatsangin** [serenity-is/Serenity#6247] [Help] [answered]
Hello dear members,
Can anyone tell me how to write the following tenant code for setting TenantId for new users in Serene .net core because the following code is giving error in serene .net core.
`protected override void SetInternalFields()
{
    base.SetInternalFields();

    if (IsCreate)
    {
        Row.Source = "site";
        Row.IsActive = Row.IsActive ?? 1;
        if (!Authorization.HasPermission(Administration.PermissionKeys.Tenants) ||
            Row.TenantId == null)
        {
            Row.TenantId = ((UserDefinition)Authorization.UserDefinition)
                .TenantId;
        }
    }

    if (IsCreate || !Row.Password.IsEmptyOrNull())
    {
        string salt = null;
        Row.PasswordHash = GenerateHash(password, ref salt);
        Row.PasswordSalt = salt;
    }
}`

I have written the above code as follows but only problem is how i can set the current logged in user tenant id in the following code:
`   base.SetInternalFields();

                if (IsCreate)
                {
                    Row.Source = "site";
                    Row.IsActive = Row.IsActive ?? 1;
                    if (!Permissions.HasPermission(PermissionKeys.Tenants) ||
                        Row.TenantId == null)
                    {
                        Row.TenantId = ((UserDefinition)Authorization.UserDefinition)
                .TenantId;

                    }
                }`
how to write  Row.TenantId = ((UserDefinition)Authorization.UserDefinition)
                .TenantId; for serene .Net Core, because it is giving an error?

**Comment by go2ismail** [comment, marked as answer]
Hi,

I'm assuming you have read through all the tutorial of multitenancy at this link: https://serenity.is/docs/tutorials/multi_tenancy/multi_tenancy . and yes, latest version serenity access UserDefinition is slightly different and become: (what its work for me)

```
        public void OnSetInternalFields(ISaveRequestHandler handler)
        {
            var username = UserAccessor.User?.Identity?.Name;
            if (UserRetrieveService.ByUsername(username) is UserDefinition user)
            {
                if (handler.IsCreate)
                    fldTenantId[handler.Row] = user.TenantId;
            }
        }
```

note:
 you have to pass the needed service via constructor:

```
        private Int32Field fldTenantId;
        protected IUserRetrieveService UserRetrieveService { get; }
        protected IUserAccessor UserAccessor { get; }
        protected IPermissionService Permissions { get; }
        protected ITextLocalizer Localizer { get; }

        public MultiTenantBehavior(IUserRetrieveService userRetrieveService, IUserAccessor userAccessor, IPermissionService permissions, ITextLocalizer localizer)
        {
            UserRetrieveService = userRetrieveService ?? throw new ArgumentNullException(nameof(userRetrieveService));
            UserAccessor = userAccessor ?? throw new ArgumentNullException(nameof(userAccessor));
            Permissions = permissions ?? throw new ArgumentNullException(nameof(permissions));
            Localizer = localizer ?? throw new ArgumentNullException(nameof(localizer));
        }
```

**Reply by imhemayatsangin** [reply]
Thank you so much dear @go2ismail
